### PR TITLE
Enhancement #16 - Remote AllowList Retrieval

### PR DESCRIPTION
**Pull request by mpeters**

### CODE REVIEW

**Review by lkatalin — APPROVED** [review]
This seems well thought-out to me - as it's the first enhancement proposal I'm reading, it's also a nice, clear example of what these look like.

**Review by lukehinds — COMMENTED** [review]
This is looking really good and very well written, a couple of things it would be good to add:

* Validation, perform some sort of validation within the tenant to ensure correct formatting and all fields are present. If not, reject it before posting to the verifier.

* Would be good to get some details about changes made to the verifier and for example would we have backwards compatibility with the old list format? For this I would say if its easily achieved, we can include that, if it results in a less than optimal approach I am ok with dropping the old standard and we release this feature as a non backwards -compat release.

File: 16_remote_allowlist_retrieval.md [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+things like accept invalid TLS certificates.
+
+* We will need to thoroughly review our handling of the cryptographic
+signatures and checksums
+
+## Design Details
+
+Propsed JSON spec:
+
+{
```
Comment (ashcrow) [inline 16_remote_allowlist_retrieval.md:189]: There are a few minor mistakes in the example spec and this mistake is on me :smile:.

- 1 missing comma
- 1 extra comma
- Missing `}`
- optional second hash example shouldn't be the same hash function

The JSON should be:

```
{
    "meta": {
        "version": "",        # Hash format version to allow for future changes if needed
        "generator": "",     # What generated the hash payload
        "timestamp": ""     #  When the hash payload was generated
    },
    "release": "",            # Release Identifier. Specifics here may be different for different compound components
    "hashes": {
        "/filesystem/path": [
            {"sha256": "HASH"}, # an object so other hashes could go side by side if needed
            {"sha128": "HASH"}  # other hashes if required/wanted
        ]
    }
}
```

**Comment by ashcrow** [comment]
> Would be good to get some details about changes made to the verifier and for example would we have backwards compatibility with the old list format? For this I would say if its easily achieved, we can include that, if it results in a less than optimal approach I am ok with dropping the old standard and we release this feature as a non backwards -compat release.

Are you thinking of being able to convert from the old format to the new format in a safe yet hands off fashion? IE: "Grab the file, oh it's the old format, verify it first then convert and use it"? Or more so allow for converting with a utility by a cluster admin/system admin?

File: 16_remote_allowlist_retrieval.md (hunk 1 above)
Comment (font) [inline 16_remote_allowlist_retrieval.md:189]: Can we surround this in a markdown code block so it's nicely formatted?

File: 16_remote_allowlist_retrieval.md [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+  * `allowlist-sig-url`
+    An HTTP URL that points to the cryptographic signature of the JSON
+    data file. Again, This URL must be secure (HTTPS), and non-redirecting
+    (no 30X HTTP codes) to reduce the chance of any MITM attacks. This
+    option cannot be specified if `allowlist-url` isn't also present. And
+    you must include either this option or `allowlist-sig` or `allowlist-checksum`
+
+  * `allowlist-sig`
+    The cryptographic signature of the remote JSON data file. This
+    option This option cannot be specified if `allowlist-url`
```
Comment (font) [inline 16_remote_allowlist_retrieval.md:134]: s/This option//

**Comment by mpeters** [comment]
> * Validation, perform some sort of validation within the tenant to ensure correct formatting and all fields are present. If not, reject it before posting to the verifier.

Good point. Adding.

> * Would be good to get some details about changes made to the verifier and for example would we have backwards compatibility with the old list format? For this I would say if its easily achieved, we can include that, if it results in a less than optimal approach I am ok with dropping the old standard and we release this feature as a non backwards -compat release.

I mention the backwards compat in the "Risks and Mitigations" sections. But I'll expand that to include mitigation strategies if backcompat can't be made to work.

**Review by lukehinds — APPROVED** [review]

**Comment by ashcrow** [comment]
Is there anything else holding this back from merging?

**Pull request merged**